smack-websocket/src/main/java/org/jivesoftware/smack/websocket/WebsocketConnectionAttemptState.java

```diff
@@ -0,0 +1,83 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jivesoftware.smack.websocket;
+
+import java.io.IOException;
+import java.net.InetAddress;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+import org.jivesoftware.smack.SmackException;
+import org.jivesoftware.smack.SmackException.EndpointConnectionException;
+import org.jivesoftware.smack.XMPPException;
+import org.jivesoftware.smack.c2s.internal.ModularXmppClientToServerConnectionInternal;
+import org.jivesoftware.smack.fsm.ConnectionStateEvent;
+import org.jivesoftware.smack.fsm.State;
+import org.jivesoftware.smack.fsm.StateTransitionResult;
+import org.jivesoftware.smack.util.rce.RemoteConnectionException;
+import org.jivesoftware.smack.websocket.XmppWebsocketTransportModule.EstablishingWebsocketConnectionState;
+import org.jivesoftware.smack.websocket.implementations.OkHttpWebsocketImpl;
+import org.jivesoftware.smack.websocket.implementations.WebsocketImplProvider;
+import org.jivesoftware.smack.websocket.implementations.WebsocketInterface;
+import org.jivesoftware.smack.websocket.rce.WebsocketRemoteConnectionEndpoint;
+
+public final class WebsocketConnectionAttemptState {
+
+    private final ModularXmppClientToServerConnectionInternal connectionInternal;
+    private final XmppWebsocketTransportModule.XmppWebsocketTransport.DiscoveredWebsocketEndpoints discoveredEndpoints;
+
+    final List<RemoteConnectionException<?>> connectionExceptions;
+    EndpointConnectionException connectionException;
+
+    final Iterator<WebsocketRemoteConnectionEndpoint> connectionEndpointIterator;
+    WebsocketRemoteConnectionEndpoint connectionEndpoint;
+    Iterator<? extends InetAddress> inetAddressIterator;
+
+    WebsocketInterface websocket;
+
+    WebsocketConnectionAttemptState(ModularXmppClientToServerConnectionInternal connectionInternal,
+                    XmppWebsocketTransportModule.XmppWebsocketTransport.DiscoveredWebsocketEndpoints discoveredWebsocketEndpoints,
+                    EstablishingWebsocketConnectionState establishingWebsocketConnectionState) throws IOException {
+        this.connectionInternal = connectionInternal;
+        this.discoveredEndpoints = discoveredWebsocketEndpoints;
+
+        List<WebsocketRemoteConnectionEndpoint> endpoints = discoveredWebsocketEndpoints.result.discoveredRemoteConnectionEndpoints;
+        connectionEndpointIterator = endpoints.iterator();
+        connectionEndpoint = connectionEndpointIterator.next();
+        connectionExceptions = new ArrayList<>(endpoints.size());
+    }
+
+    StateTransitionResult.Failure establishWebsocketConnection() throws InterruptedException {
+
+        try {
+            websocket = WebsocketImplProvider.getWebsocketImpl(OkHttpWebsocketImpl.class, connectionInternal, discoveredEndpoints);
```

Comment:
Never group methods throwing disjunct exceptions in the same try-catch block! Use multiple try-catch blocks instead!